webbeans-impl/src/main/java/org/apache/webbeans/component/AbstractOwbBean.java

```diff
@@ -88,6 +92,8 @@ protected AbstractOwbBean(WebBeansContext webBeansContext,
         this.webBeansType = webBeansType;
         this.beanClass = beanClass;
         this.webBeansContext = webBeansContext;
+
+        returnType = (Class<T>) WebBeansUtil.resolveReturnType(beanAttributes.getTypes());
```

Comment:
this looks neat but is kind of against the spec which explicit states what `beanClass` is:

> getBeanClass() returns the bean class of the managed bean or of the bean that declares the producer method or field.

So for me there is no ambiguity but I ack some extensions are buggy and don't comply to CDI (using extension class or Object, they should define it to the main bean class type - potentially without generics)

Reply:
Hmm not sure if synthetic beans are managed beans: https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0#implementation
> Portable extensions and build compatible extensions may provide other kinds of beans.

This implies that beans added via a portable extension are not necessarily managed beans -> the restriction in the javadoc does not necessarily apply IMO which means OWB can't rely on beanClass to be something meaningful or even represent the type of the Bean (which is the case here)


Even if that is not the case and beans registered via a portable extension are indeed managed.. Is there a requirement for OWB to use beanClass to determine which type to proxy? (At least can't find something in the specs quickly searching through it)
Looking for the most specific [type](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0#bean_types) in getTypes _should_ also be sufficient I think

But I could also be completely wrong here and this discussion really should happen on the Mojarra/Weld side 😉

Reply:
Not sure thre is much ambiguity (for ex 5.1.4 speaks of `Bean` whatever the source is).
So in CDI you have Beans, some specificities in the spec about producers (very old weld limitations) and be it.
Synthetic beans are just `Bean` at the end and behave the same.
You would note that the extension has a specific API (getSource) so `beanClass` is literally intended to behave as a managed bean.

> Is there a requirement for OWB to use beanClass to determine which type to proxy.

Requirement as "in the spec" no, as "in the impl" yes since types can not overlap enough and you can pick the wrong one if you don't use bean class.

> Looking for the most specific [type](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0#bean_types) in getTypes should also be sufficient I think

Is there any requirement? :angel: 
But this has some issues since you can not use a large enough type so will lead to broken cases as soon as types != union(typesOf(beanClass)) for a "standard"/simple bean case.